keras_unet/models/custom_unet.py

```diff
@@ -1,75 +1,113 @@
 from keras.models import Model
-from keras.layers import BatchNormalization, Conv2D, Conv2DTranspose, MaxPooling2D, Dropout, UpSampling2D, Input, concatenate
+from keras.layers import (
+    BatchNormalization,
+    Conv2D,
+    Conv2DTranspose,
+    MaxPooling2D,
+    SpatialDropout2D,
+    UpSampling2D,
+    Input,
+    concatenate,
+    Activation,
+)
+
 
 def upsample_conv(filters, kernel_size, strides, padding):
     return Conv2DTranspose(filters, kernel_size, strides=strides, padding=padding)
 
+
 def upsample_simple(filters, kernel_size, strides, padding):
     return UpSampling2D(strides)
 
+
 def conv2d_block(
-    inputs, 
-    use_batch_norm=True, 
-    dropout=0.3, 
-    filters=16, 
-    kernel_size=(3,3), 
-    activation='relu', 
-    kernel_initializer='he_normal', 
-    padding='same'):
-    
-    c = Conv2D(filters, kernel_size, activation=activation, kernel_initializer=kernel_initializer, padding=padding) (inputs)
+    inputs,
+    use_batch_norm=True,
+    dropout=0.3,
+    filters=16,
+    kernel_size=(3, 3),
+    activation="relu",
+    kernel_initializer="he_normal",
+    padding="same",
+):
+
+    c = Conv2D(
+        filters,
+        kernel_size,
+        activation=None,
+        kernel_initializer=kernel_initializer,
+        padding=padding,
+        use_bias=not use_batch_norm,
+    )(inputs)
     if use_batch_norm:
         c = BatchNormalization()(c)
+        c = Activation(activation)(c)
     if dropout > 0.0:
-        c = Dropout(dropout)(c)
-    c = Conv2D(filters, kernel_size, activation=activation, kernel_initializer=kernel_initializer, padding=padding) (c)
+        c = SpatialDropout2D(dropout)(c)
+    c = Conv2D(
+        filters,
+        kernel_size,
+        activation=None,
+        kernel_initializer=kernel_initializer,
+        padding=padding,
+        use_bias=not use_batch_norm,
```

Comment:
Batch Norm includes a bias parameter that just rescales the bias from the previous layer to whatever is needed. It's not problematic to use bias with batch norm, but it's unnecessary since the BN will do what it wants anyway and it reduces performance a little.

Reply:
Thanks for explaination! Makes perfect sense